### PR TITLE
HIVE-23768: Metastore's update service wrongly strips partition column stats from the cache

### PR DESCRIPTION
**Pull request by zabetak**

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 1]
```diff
@@ -851,8 +851,6 @@ private void updateTableColStats(RawStore rawStore, String catName, String dbNam
             sharedCache.refreshTableColStatsInCache(StringUtils.normalizeIdentifier(catName),
                 StringUtils.normalizeIdentifier(dbName), StringUtils.normalizeIdentifier(tblName),
                 tableColStats.getStatsObj());
-            // Update the table to get consistent stats state.
-            sharedCache.alterTableInCache(catName, dbName, tblName, table);
```
Comment (kishendas) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:855]: Sorry, I am bit confused looking at this diff. So, the original issue seems to be - "Metastore's update service wrongly strips partition column stats from the cache in an attempt to update them." . How are we fixing this issue by not updating the stats in the cache ? Wouldn't the right fix for this is to to ensure sharedCache.alterTableInCache and sharedCache.alterPartitionInCache methods do the right thing and not incorrectly remove partition column stats ?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 2]
```diff
@@ -900,13 +898,6 @@ private void updateTablePartitionColStats(RawStore rawStore, String catName, Str
               rawStore.getPartitionColumnStatistics(catName, dbName, tblName, partNames, colNames, CacheUtils.HIVE_ENGINE);
           Deadline.stopTimer();
           sharedCache.refreshPartitionColStatsInCache(catName, dbName, tblName, partitionColStats);
-          Deadline.startTimer("getPartitionsByNames");
-          List<Partition> parts = rawStore.getPartitionsByNames(catName, dbName, tblName, partNames);
-          Deadline.stopTimer();
-          // Also save partitions for consistency as they have the stats state.
-          for (Partition part : parts) {
-            sharedCache.alterPartitionInCache(catName, dbName, tblName, part.getValues(), part);
```
Comment (kishendas) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:908]: Same concern here as previous. How do we fix this issue by not updating the cache at all ?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 3]
```diff
@@ -2264,6 +2255,8 @@ private MergedColumnStatsForPartitions mergeColStatsForPartitions(String catName
       if (colStatsMap.size() < 1) {
         LOG.debug("No stats data found for: dbName={} tblName= {} partNames= {} colNames= ", dbName, tblName, partNames,
             colNames);
+        // TODO: If we don't find any stats then most likely we should return null. Returning an empty object will not
+        // trigger the lookup in the raw store and we will end up with missing stats.
```
Comment (kishendas) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2259]: Please create a JIRA for this TODO and add a reference in the comment.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 2 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:908]: The explanation is hidden in the commit message :)

Using alterPartitionInCache is problematic:
1. It empties all relevant stats about partitions even those updated just now.
2. It is not necessary since partitions were updated/refreshed just before calling the method updateTablePartitionColStats via [updateTablePartitions](https://github.com/apache/hive/blob/31ee14644bf6105360d6266baa8c6c8060d38ea3/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java#L870).
3. It is not meant to be used for syncing the cache but rather handle updates in partitions.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 1 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:855]: If my understanding is correct the cache is already updated by the time we arrive here. Check my comment regarding `alterPartitionsInCache` and if something is not clear I can elaborate more.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 2 above)
Comment (kishendas) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:908]: Ok, thanks for the explanation. Would you need below lines, since you no longer use List<Partition>  ?

          Deadline.startTimer("getPartitionsByNames");
          List<Partition> parts = rawStore.getPartitionsByNames(catName, dbName, tblName, partNames);

Comment (zabetak) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:908]: They are already removed aren't they? Maybe, I didn't understand what you mean.

Comment (kishendas) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:908]: Ok, in review mode, only last two lines appear. I can see that they are removed in "Files changed" tab.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 3 above)
Comment (kishendas) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2259]: Did you create a JIRA for this ?

Comment (zabetak) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2259]: Yeap, check https://issues.apache.org/jira/browse/HIVE-23781. I was planning to push a new commit now to update the comment.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java [hunk 4]
```diff
@@ -136,17 +129,19 @@
...
-    Deadline.startTimer("");
-    objectStore.dropPartitions(DEFAULT_CATALOG_NAME, db1Ptbl1.getDbName(), db1Ptbl1.getTableName(), db1Ptbl1PtnNames);
-    objectStore.dropTable(DEFAULT_CATALOG_NAME, db1Ptbl1.getDbName(), db1Ptbl1.getTableName());
-    objectStore.dropTable(DEFAULT_CATALOG_NAME, db2Utbl1.getDbName(), db2Utbl1.getTableName());
-    Deadline.startTimer("");
-    objectStore.dropPartitions(DEFAULT_CATALOG_NAME, db2Ptbl1.getDbName(), db2Ptbl1.getTableName(), db2Ptbl1PtnNames);
-    objectStore.dropTable(DEFAULT_CATALOG_NAME, db2Ptbl1.getDbName(), db2Ptbl1.getTableName());
-    objectStore.dropDatabase(DEFAULT_CATALOG_NAME, db1.getName());
-    objectStore.dropDatabase(DEFAULT_CATALOG_NAME, db2.getName());
+    for (String clg : objectStore.getCatalogs()) {
```
Comment (kishendas) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:132]: Why is this change required ? I think in the test case, it's preferable to explicitly add drop statements , so that we know what's going on.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 2 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:908]: Indeed some times its confusing :)

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java [hunk 5]
```diff
@@ -1566,160 +1643,6 @@ private Table createUnpartitionedTableObject(Database db) {
     return tbl;
   }
 
-  private TableAndColStats createUnpartitionedTableObjectWithColStats(Database db) {
```
Comment (kishendas) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:1569]: Why is this test case being removed ?

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java (hunk 4 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:132]: The `setUp` method adds things in the Metastore that are not necessarily needed by every test and this was the case even before my changes. In the new test, I added some more tables/columns cause I think it made things easier to understand at least for me. If I was to follow the current pattern I would have to add the tables, cols, partitions, etc., in the `setUp` method and then drop them in the `teardown` method but then they would be visible to all other tests as well. 

I like the philosophy where unit tests are minimal (least possible state) thus I would prefer if every test starts with a clean Metastore and ends with a clean Metastore. I didn't change the `setUp` in order not to introduce too many unrelated changes but I made one step forward in the `teardown`.  

If every test in this class was using the same state then I would definitely agree that the `setUp` and `teardown` methods should be the way they are right now.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java (hunk 5 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:1569]: This is not a test case it is just a private unused method.

Comment (kishendas) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:1569]: Ok, thanks for the cleanup.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java (hunk 4 above)
Comment (kishendas) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:132]: Sounds good.
